core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java

```diff
@@ -82,28 +82,48 @@ public Schema record(Schema record, List<String> names, Iterable<Schema.Field> s
     List<Schema.Field> updatedFields = Lists.newArrayListWithExpectedSize(struct.fields().size());
     List<Types.NestedField> expectedFields = struct.fields();
     for (int i = 0; i < expectedFields.size(); i += 1) {
-      Types.NestedField field = expectedFields.get(i);
+      Types.NestedField expectedField = expectedFields.get(i);
 
       // detect reordering
-      if (i < fields.size() && !field.name().equals(fields.get(i).name())) {
+      if (i < recordFields.size() && !expectedField.name().equals(recordFields.get(i).name())) {
         hasChange = true;
       }
 
-      Schema.Field avroField = updateMap.get(AvroSchemaUtil.makeCompatibleName(field.name()));
+      Schema.Field avroField = updateMap.get(AvroSchemaUtil.makeCompatibleName(expectedField.name()));
 
       if (avroField != null) {
-        updatedFields.add(avroField);
-
+        if (expectedField.hasDefaultValue() && AvroSchemaUtil.isNonNullDefault(expectedField.getDefaultValue()) &&
```

Comment:
Could you explain the purpose of this `if` branch in a comment in the code? Also, adding a comment in the `else` branch would help. Making sure explaining the `(!avroField.hasDefaultValue() || avroField.defaultVal() == JsonProperties.NULL_VALUE))` part would be helpful.

Reply:
done